### PR TITLE
doc/dev/cephadm: document the cephadm exporter design

### PR DESCRIPTION
**Pull request by pcuzner**
Development document defining the goals, requirements,
design and implementation details for the exporter mode
of the cephadm 'binary'

Signed-off-by: Paul Cuzner <pcuzner@redhat.com>

### CODE REVIEW

**Review by epuertat — APPROVED** [review]
Looks good @pcuzner !

File: doc/dev/cephadm/cephadm-exporter.rst [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+.. note:: If the service is already deployed any attempt to change or clear the configuration will
+    be denied. In order to change settings you must remove the service, apply the required configuration
+    and re-apply (``ceph orch apply cephadm-exporter``)
+
+
+
+New Ceph Configuration Keys
+===========================
+The exporter configuration is persisted to the monitor's KV store, as the following settings;
```
Comment (zdover23) [inline doc/dev/cephadm/cephadm-exporter.rst:136]:
```suggestion
The exporter configuration is persisted to the monitor's KV store, as the following settings:
```

File: doc/dev/cephadm/cephadm-exporter.rst [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        if r.status in [200, 206]:
+
+            raw=r.read()  # bytes string
+            js=json.loads(raw.decode())
+            print(json.dumps(js, indent=2))
+        elapsed = time.time() - s_time
+        print(f"Elapsed secs : {elapsed}")
+
+
+.. note:: the above example is using python3, and assumes that you've extracted the config using the get-exporter-config command.
```
Comment (zdover23) [inline doc/dev/cephadm/cephadm-exporter.rst:244]:
```suggestion
.. note:: the above example uses python3 and assumes that you've extracted the config using the ``get-exporter-config`` command.
```

File: doc/dev/cephadm/cephadm-exporter.rst [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+
+Implementation Specific Details
+===============================
+
+Like a typical container based deployment, the exporter is deployed to a directory under ``/var/lib/ceph/<fsid>``. The
```
Comment (zdover23) [inline doc/dev/cephadm/cephadm-exporter.rst:250]:
```suggestion
The exporter is deployed to a directory under ``/var/lib/ceph/<fsid>``, as in a typical container-based deployment. The 
```

File: doc/dev/cephadm/cephadm-exporter.rst (hunk 2 above)
Comment (pcuzner) [reply doc/dev/cephadm/cephadm-exporter.rst:244]: ack

File: doc/dev/cephadm/cephadm-exporter.rst (hunk 3 above)
Comment (pcuzner) [reply doc/dev/cephadm/cephadm-exporter.rst:250]: re-phrased. thanks

**Comment by zdover23** [comment]
jenkins make check

**Comment by zdover23** [comment]
jenkins test make check

**Comment by pcuzner** [comment]
@sebastian-philipp would it be possible to get some review from orchestrators group please?

File: doc/dev/cephadm/cephadm-exporter.rst [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+.. csv-table:: Supported HTTP Responses
+    :header: "Status Code", "Meaning"
+
+    "200", "OK"
+    "204", "the thread associated with this request is no longer active, no data is returned"
+    "206", "some threads have stopped, so some content is missing"
+    "401", "request is not authorised - check your token is correct"
+    "404", "URL is malformed, not found"
+    "500", "all threads have stopped - unable to provide any metadata for the host"
```
Comment (sebastian-philipp) [inline doc/dev/cephadm/cephadm-exporter.rst:172]: Isn't this already documented by saying we're using a REST API? Sounds duplicated to list standard HTTP codes here.

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
jenkins test api

**Comment by sebastian-philipp** [comment]
jenkins test api

**Comment by sebastian-philipp** [comment]
needs rebase

File: doc/dev/cephadm/cephadm-exporter.rst (hunk 4 above)
Comment (pcuzner) [reply doc/dev/cephadm/cephadm-exporter.rst:172]: yeah - the codes are standard, as you'd expect...but what I'm documenting here is the meaning wrt how the exporter is working or perhaps not working :)

**Pull request merged**